src/services/refactors/moveToNewFile.ts

```diff
@@ -0,0 +1,359 @@
+        ];
+    }
+
+    function deleteUnusedOldImports(oldFile: SourceFile, toMove: ReadonlyArray<Statement>, changes: textChanges.ChangeTracker, toDelete: ReadonlySymbolSet, checker: TypeChecker) {
+        for (const statement of oldFile.statements) {
+            if (!contains(toMove, statement) && isImportDeclaration(statement)) {
+                deleteUnusedImports(oldFile, statement, changes, name => toDelete.has(checker.getSymbolAtLocation(name)));
+            }
+        }
+    }
+
+    function updateImportsInOtherFiles(changes: textChanges.ChangeTracker, program: Program, oldFile: SourceFile, movedSymbols: ReadonlySymbolSet, newModuleName: string): void {
+        const checker = program.getTypeChecker();
+        for (const sourceFile of program.getSourceFiles()) {
+            if (sourceFile === oldFile) continue;
+            for (const statement of sourceFile.statements) {
+                if (!isImportDeclaration(statement) || !isStringLiteral(statement.moduleSpecifier)) continue;
+
+                const shouldMove = (name: Identifier): boolean => movedSymbols.has(skipAlias(checker.getSymbolAtLocation(name), checker));
+                deleteUnusedImports(sourceFile, statement, changes, shouldMove);
+                const newModuleSpecifier = combinePaths(getDirectoryPath(statement.moduleSpecifier.text), newModuleName);
+                const newImportDeclaration = filterImport(statement, createLiteral(newModuleSpecifier), shouldMove);
+                if (newImportDeclaration) changes.insertNodeAfter(sourceFile, statement, newImportDeclaration);
+            }
+        }
+    }
+
+    function createOldFileImportsFromNewFile(newFileNeedExport: ReadonlySymbolSet, newFileNameWithExtension: string): ImportDeclaration | undefined {
+        let defaultImport: Identifier | undefined;
+        const imports: ImportSpecifier[] = [];
+        newFileNeedExport.forEach(symbol => {
+            if (symbol.escapedName === InternalSymbolName.Default) {
+                defaultImport = createIdentifier(symbolNameNoDefault(symbol));
+            }
+            else {
+                imports.push(createImportSpecifier(undefined, createIdentifier(symbol.name)));
+            }
+        });
+        return makeImportIfNecessary(defaultImport, imports, ensurePathIsRelative(newFileNameWithExtension));
+    }
+
+    function addExports(toMove: ReadonlyArray<Statement>, needExport: ReadonlySymbolSet): ReadonlyArray<Statement> {
+        return toMove.map(statement => {
+            return !hasModifier(statement, ModifierFlags.Export) && forEachTopLevelDeclaration(statement, d => needExport.has(Debug.assertDefined(d.symbol)))
+                ? addExport(statement as TopLevelDeclarationStatement)
+                : statement;
+        });
+    }
+
+    function deleteUnusedImports(sourceFile: SourceFile, importDecl: ImportDeclaration, changes: textChanges.ChangeTracker, isUnused: (name: Identifier) => boolean): void {

```

Comment:
this looks a lot like the logic we use in `fixUnusedIdentifer` for module imports.. can we consolidate?

Reply:
I don't think that's easy since `fixUnusedIdentifier` deletes a single identifier starting from the identifier, but here we have a set of symbols and want to recurse down starting from the sourcefile and delete everything in the set.